### DISCUSSION TITLE
chiseltest with mill as the building tool

### DISCUSSION THREAD

**Discussion by wangpengabc** [chipsalliance/chisel#1988] [Q&A] [answered]
here is my project：
https://github.com/wangpengabc/systolic

I build this project from https://github.com/OpenXiangShan/chisel-playground which uses utest as test tool. I want to change the testing tool from utest to chiseltest(known as chisel tester2), and the building tool is mill. The tested class is in https://github.com/wangpengabc/systolic/tree/master/systolic/test/src, when i run `make test` , the output is 
```
$ make test
mill -i __.test
[51/60] systolic.test.compile 
[info] compiling 1 Scala source to /home/wangpeng/riscv/chisel/systolic/git_pro/chisel_systolic/systolic_wp/out/systolic/test/compile/dest/classes ...
[info] done compiling
[60/60] systolic.test.test
```

when i run `mill systolic.test._`, the output is 
```
$ mill systolic.test._
[51/88] systolic.test.resolvedAmmoniteReplIvyDeps | Downloading [1/1] artifacts [51/88] systolic.test.resolvedAmmoniteReplIvyDeps | Downloading [1/1] artifacts [51/88] systolic.test.resolvedAmmoniteReplIvyDeps | Downloading [2/2] artifacts [51/88] systolic.test.resolvedAmmoniteReplIvyDeps | Downloading [2/2] artifacts (~0/0 bytes)
1 targets failed
systolic.test.resolvedAmmoniteReplIvyDeps 
Resolution failed for 1 modules:
--------------------------------------------
  com.lihaoyi:ammonite_2.12.13:2.3.8-65-0f0d597f 
	not found: /home/wangpeng/.ivy2/local/com.lihaoyi/ammonite_2.12.13/2.3.8-65-0f0d597f/ivys/ivy.xml
	not found: https://repo1.maven.org/maven2/com/lihaoyi/ammonite_2.12.13/2.3.8-65-0f0d597f/ammonite_2.12.13-2.3.8-65-0f0d597f.pom
```

**Comment by sequencer** [comment, marked as answer]
This has been tested successfully on my Mac:
```
❯ make test
mill -i __.test
Compiling /private/tmp/systolic/build.sc
[31/60] systolic.compile uginClasspath | Downloading [1/1] artifacts (~1925/1925Compiling compiler interface...
[info] compiling 5 Scala sources to /private/tmp/systolic/out/systolic/compile/dest/classes ...
[info] done compiling
[51/60] systolic.test.compile 
[info] compiling 1 Scala source to /private/tmp/systolic/out/systolic/test/compile/dest/classes ...
[info] done compiling
[60/60] systolic.test.test 
```
with mill:
```
Mill Build Tool version 0.9.8
Java version: 17, vendor: Homebrew, runtime: /opt/homebrew/Cellar/openjdk/16.0.1/libexec/openjdk.jdk/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "Mac OS X", version: 11.4, arch: aarch64
```
I think you should blame your network connection, fix it and try again.

**Reply by wangpengabc** [reply]
Thanks! But on my local computer, when i run "make test", it is ok without any problem, the output is the same as yours. But i dont think the testcase named systolicTest runs, because the program "Console.printf("something")" does not run and the console does not print the strings.
![image](https://user-images.githubusercontent.com/22144608/123753220-2f288c00-d8ec-11eb-9ff0-c262655f64a6.png)

**Reply by sequencer** [reply]
You are using utest as framework from `build.sc`:
https://github.com/wangpengabc/systolic/blob/cca81e7ab079fcec84e9634cc6dafab92494acc5/build.sc#L27

While actually you are writing a ScalaTest framework.

**Reply by wangpengabc** [reply]
Thanks! The problem is solved! I adjust the build.sc, change the test ivyDeps's part:
```
  object test extends Tests with TestModule.ScalaTest{
    override def ivyDeps = m.ivyDeps() ++ Agg(
      ivy"org.scalatest::scalatest:3.2.9",
      ivy"com.lihaoyi::utest:0.7.10",
      ivy"edu.berkeley.cs::chiseltest:0.3.3",
    )

  }
```

I add the scalatest package. My project works! 

PS: there is an error when i peek a Vec, I think chisel tester2 maybe do not support peek Vec? I have not figured out.
![image](https://user-images.githubusercontent.com/22144608/123778624-de249200-d903-11eb-9830-acc977d9ff8d.png)

**Reply by sequencer** [reply]
IIRC, this should be able to work with https://github.com/ucb-bar/chisel-testers2/pull/280

**Reply by wangpengabc** [reply]
The version of chiseltest in my project is 0.3.3, and the publish date is 1, April.  The date of fixing Vec peeking is 29,April , so i think this function is not coporated in chiseltest 0.3.3.

**Reply by sequencer** [reply]
Try 0.5 snapshot or build from source via https://github.com/sequencer/chipsalliance-playground

**Reply by carlosedp** [reply]
I keep a giter8 template for my Chisel projects and it has both SBT and Mill build definitions with updated deps.
Take a look at: https://github.com/carlosedp/chisel-template.g8

**Reply by sequencer** [reply]
Thanks @carlosedp, g8 is another good try ;p
But fundamentally, I think chiseltest should have a more active version release strategy(since it has no old version backports) to avoid asking user build from source like my playground, it is only for devs not for users.

**Reply by wangpengabc** [reply]
> Try 0.5 snapshot or build from source via https://github.com/sequencer/chipsalliance-playground

Thanks！ I modified from https://github.com/sequencer/chipsalliance-playground and it succeeded.
